### PR TITLE
Hyperoptim

### PR DESCRIPTION
**Pull request by mknull**
Add hyperoptimization capabilities

### CODE REVIEW

File: hyperoptimization/explore.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
+import hpbandster.core.result as hpres
+import hpbandster.visualization as hpvis
+import json
+import matplotlib.pyplot as plt
+from collections import Counter
+import numpy as np
+import os
+
+
+class ValidFreeEnergy:
```
Comment (jdrefs) [inline hyperoptimization/explore.py:10]: For a general hyperparameter optimization framework, it may be desirable to provide an interface that allows to implement different optimization metrics (including free energies) as generically as possible.

File: hyperoptimization/run_best_config.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
+import hpbandster.core.result as hpres
+from hyperoptimization.runs import from_config as run
+from hyperoptimization.workers import TVAEWorker
```
Comment (jdrefs) [inline hyperoptimization/run_best_config.py:3]: Do I understand correctly that this method is tailored to a particular model?

File: hyperoptimization/run_best_config.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    parser.add_argument("--Ksize", type=int, default=3, help="size of each K^n set")
+    parser.add_argument("--epochs", type=int, default=40, help="number of training epochs")
+    parser.add_argument(
+        "--net-shape",
+        required=True,
+        type=parse_net_shape,
+        help="column-separated list of layer sizes",
+    )
+    parser.add_argument("--min_lr", type=float, help="MLP min learning rate", required=True)
+    parser.add_argument("--max_lr", type=float, help="MLP max learning rate", required=True)
```
Comment (jdrefs) [inline hyperoptimization/run_best_config.py:22]: Do I understand correctly that this method is tailored to a particular model? (similar comment above)

File: hyperoptimization/run_best_config.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+
+def parse_net_shape(net_shape: str) -> Tuple[int, ...]:
+    """
+    Parse string with TVAE shape into a tuple.
+
+    :param net_shape: column-separated list of integers, e.g. `"10:10:2"`
+    :returns: a tuple with the shape as integers, e.g. `(10,10,2)`
+    """
+    return tuple(map(int, net_shape.split(":")))
```
Comment (jdrefs) [inline hyperoptimization/run_best_config.py:40]: Do I understand correctly that this method is tailored to a particular model? (similar comment above)

File: hyperoptimization/run_best_config.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+
+parser = experiment(hyperopt(Parser()))
+parsed_args = parser.parse_args()
+
+worker = TVAEWorker
```
Comment (jdrefs) [inline hyperoptimization/run_best_config.py:46]: Do I understand correctly that this method is tailored to a particular model? (similar comment above)

File: hyperoptimization/run_best_config.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+path = ""
+result = hpres.logged_results_to_HBS_result(path)
+all_runs = result.get_all_runs()
+id2conf = result.get_id2config_mapping()
+
+ordered_by_loss = sorted_by_value(all_runs, key="loss")
```
Comment (jdrefs) [inline hyperoptimization/run_best_config.py:53]: Do I understand correctly that this method is tailored to a particular optimization objective?

File: hyperoptimization/workers.py [hunk 7]
```diff
@@ -0,0 +1,627 @@
...
+
+        return {
+            "loss": -train_F if -train_F else np.nan,
+            "info": {
+                "test accuracy": test_F,
+                "train accuracy": train_F,
+                "validation accuracy": valid_F,
+                "number of parameters": model._external_model.number_of_parameters(),
+            },
+        }
```
Comment (jdrefs) [inline hyperoptimization/workers.py:202]: Do I understand correctly that this implies dependency on a particular model and optimization metric? (similar comment above)

**Review by jdrefs — COMMENTED** [review]
Afaics, this MR provides an extensive implementation to run hyperparameter optimization (HPO) for TVAE on bars test using free energies as HPO objective. As significant parts of the code are tailored to this application, i.e., tailored to a concrete model and a concrete HPO objective (e.g., in `neural_models.py`, `test_neural_models.py`, `workers.py`; also see comments), my suggestion wouId be to make this repository part of the [`examples` ](https://github.com/tvlearn/tvo/tree/master/examples). Future refactoring may aim at bundling all methods that are generically applicable to different models and experiments and make them part of `tvo.utils`.

File: tvo/models/__init__.py [hunk 8]
```diff
@@ -1,8 +1,8 @@
 from .noisyor import NoisyOR
 from .bsc import BSC
-from .sssc import SSSC
 from .tvae import GaussianTVAE, BernoulliTVAE
 from .gmm import GMM
 from .pmm import PMM
+from .sssc import SSSC
```
[the remark this answers is not in the document]
Comment (mknull) [reply tvo/models/__init__.py:6]: Not required.

File: hyperoptimization/workers.py [hunk 9]
```diff
@@ -0,0 +1,627 @@
...
+
+        try:
+            self.cyclic_lr = parsed_args.cyclic_lr
+        except AttributeError:
+            self.cyclic_lr = False
+
+        # infer hyperparameter status
+        self.is_hyperparameter_S = not (S)
+        self.is_hyperparameter_H = not (H)
+        self.is_hyperparameter_EEM = False
```
[the remark this answers is not in the document]
Comment (mknull) [reply hyperoptimization/workers.py:99]: Thanks

File: hyperoptimization/explore.py [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+    hpvis.performance_histogram_model_vs_random(all_runs, result.get_id2config_mapping(), show=True)
+
+    plt.show()
+
+
+if __name__ == "__main__":
+    # path = 'results'
+    path = os.path.abspath(
+        os.path.join(os.path.abspath(__file__), "../../dynamically_binarized_mnist/results_2")
+    )
```
[the remark this answers is not in the document]
Comment (mknull) [reply hyperoptimization/explore.py:171]: Option could be moved to CLI, but that would be premature.

File: hyperoptimization/run_best_config.py (hunk 2 above)
Comment (mknull) [reply hyperoptimization/run_best_config.py:3]: Models with different hyperparameters need to have them declared. However the method differentiates bewteen models and variational optimization schemes.

File: hyperoptimization/run_best_config.py (hunk 3 above)
Comment (mknull) [reply hyperoptimization/run_best_config.py:22]: As above, hyperparameter optimization requires different decisions for different parameters. E.g. a sparsity parameter is a continuous value between [0,1] but the learning rate as above has different rates. Similarly, booleans are discrete values that need different treatment.

File: hyperoptimization/run_best_config.py (hunk 4 above)
Comment (mknull) [reply hyperoptimization/run_best_config.py:40]: Yes, this is a function necessary to the model being run.

File: hyperoptimization/run_best_config.py (hunk 5 above)
Comment (mknull) [reply hyperoptimization/run_best_config.py:46]: As above.

File: hyperoptimization/run_best_config.py (hunk 6 above)
Comment (mknull) [reply hyperoptimization/run_best_config.py:53]: Not necessarily. E.g. if another objective is treated then the following should still work.

File: hyperoptimization/workers.py (hunk 7 above)
Comment (mknull) [reply hyperoptimization/workers.py:202]: This is for a particular worker. Individual models require individual workers.
